frame/contracts/src/lib.rs

```diff
@@ -1089,13 +1155,22 @@ impl<T: Config> Invokable<T> for InstantiateInput<T> {
 	) -> InternalOutput<T, Self::Output> {
 		let mut storage_deposit = Default::default();
 		let try_exec = || {
+			// Root origin is not allowed here.
```

Comment:
I have refactored the code so root does never reach that point. But the answer to 'why' is because root cannot 'instantiate', only 'call'
Regarding the closure, I don't know, that's part of PR [#13443](https://github.com/paritytech/substrate/pull/13443)